phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -638,7 +638,13 @@ public void preBatchMutateWithExceptions(ObserverContext<RegionCoprocessorEnviro
       }
 
       long start = EnvironmentEdgeManager.currentTimeMillis();
+      if (replayWrite == null) {
+          this.builder.scanCurrentRowStates(context.rowsToLock, indexMetaData, now);
+      }
       prepareIndexMutations(c, miniBatchOp, context, mutations, now, indexMetaData);
+      if (replayWrite == null) {
+          this.builder.removeRowStates(context.rowsToLock);
```

Comment:
If prepareIndexMutations throws an exception so that this.builder.removeRowStates never gets called, are we still ok, or does this need a finally block? Seems like the LocalTable cache of results would just fill and fill until we OOM. 

Alternately, the results cache could have an upper limit beyond which it auto-removes older items, but that's probably more complex.